chainindex/events.go

```diff
@@ -0,0 +1,501 @@
+package chainindex
+
+import (
+	"bytes"
+	"context"
+	"database/sql"
+	"fmt"
+	"math"
+	"sort"
+	"strings"
+
+	"github.com/ipfs/go-cid"
+	cbg "github.com/whyrusleeping/cbor-gen"
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/go-address"
+	amt4 "github.com/filecoin-project/go-amt-ipld/v4"
+	"github.com/filecoin-project/go-state-types/abi"
+	blockadt "github.com/filecoin-project/specs-actors/actors/util/adt"
+
+	"github.com/filecoin-project/lotus/chain/types"
+)
+
+type executedMessage struct {
+	msg types.ChainMsg
+	rct types.MessageReceipt
+	// events extracted from receipt
+	evs []types.Event
+}
+
+// events are indexed against their inclusion/message tipset when we get the corresponding execution tipset
+func (si *SqliteIndexer) indexEvents(ctx context.Context, tx *sql.Tx, msgTs *types.TipSet, executionTs *types.TipSet) error {
+	// check if we have an event indexed for any message in the `msgTs` tipset -> if so, there's nothig to do here
+	// this makes event inserts idempotent
+	msgTsKeyCidBytes, err := toTipsetKeyCidBytes(msgTs)
+	if err != nil {
+		return xerrors.Errorf("failed to get tipset key cid: %w", err)
+	}
+
+	// if we've already indexed events for this tipset, mark them as unreverted and return
+	res, err := tx.Stmt(si.updateEventsToNonRevertedStmt).ExecContext(ctx, msgTsKeyCidBytes)
+	if err != nil {
+		return xerrors.Errorf("failed to unrevert events for tipset: %w", err)
+	}
+	rows, err := res.RowsAffected()
+	if err != nil {
+		return xerrors.Errorf("failed to get rows affected by unreverting events for tipset: %w", err)
+	}
+	if rows > 0 {
+		log.Infof("unreverted %d events for tipset: %s", rows, msgTs.Key())
+		return nil
+	}
+
+	if !si.cs.IsStoringEvents() {
+		return nil
+	}
+
+	ems, err := si.loadExecutedMessages(ctx, msgTs, executionTs)
+	if err != nil {
+		return xerrors.Errorf("failed to load executed messages: %w", err)
+	}
+	eventCount := 0
+	addressLookups := make(map[abi.ActorID]address.Address)
+
+	for _, em := range ems {
+		msgCidBytes := em.msg.Cid().Bytes()
```

Comment:
I was thinking that we could opt for doing `Cid().Hash().Bytes()` for both message CID and tipset, and we'd save at least 4 bytes and a common prefix

Reply:
since we're breaking anyway

Reply:
but .. reconstructing would be mildly awkward since we need to re-insert the codec

maybe not worth the bother

Reply:
But then how do we go back from the bytes to the cid without also persisting the codec ?